lib/cosmos/dart/script/dart_import.rb

```diff
@@ -0,0 +1,166 @@
+# This code must be run on the database server
+# The file to be imported should be placed in its final storage location - Note that it is
+# imported in place with algorithms that attempt to prevent duplicate creation of
+# Database entries
+
+require 'ostruct'
+require 'optparse'
+
+options = OpenStruct.new
+options.force = false
+
+parser = OptionParser.new do |option_parser|
+  option_parser.banner = "Usage: dart_import filename"
+  option_parser.separator("")
+
+  # Create the help option
+  option_parser.on("-h", "--help", "Show this message") do
+    puts option_parser
+    exit
+  end
+
+  # Create the version option
+  option_parser.on("-v", "--version", "Show version") do
+    puts "COSMOS Version: #{COSMOS_VERSION}"
+    puts "User Version: #{USER_VERSION}" if defined? USER_VERSION
+    exit
+  end
+
+  # Create the system option
+  option_parser.on("--system FILE", "Use an alternative system.txt file") do |arg|
+    System.instance(File.join(USERPATH, 'config', 'system', arg))
+  end
+
+  # Create the force option
+  option_parser.on("-f", "--force", "Force parsing entire file") do
+    options.force = true
+  end
+end
+
+parser.parse!(ARGV)
+
+require File.expand_path('../../config/environment', __FILE__)
+require 'dart_common'
+include DartCommon
+
+Cosmos::Logger.level = Cosmos::Logger::INFO
+
+Cosmos.catch_fatal_exception do
+
+  sync_targets_and_packets()
+
+  unless ARGV[0]
+    puts parser
+    exit
+  end
+
+  filename = File.expand_path(ARGV[0])
+
+  directory = File.dirname(filename)
+  if directory != Cosmos::System.paths['DART_DATA']
+    Cosmos::Logger.fatal("Imported files must be in \"#{Cosmos::System.paths['DART_DATA']}\"")
+    Cosmos::Logger.fatal("  File is in: \"#{directory}\"")
+    exit(1)
+  end
+
+  # Make sure this file isn't already imported
+  packet_log = PacketLog.where("filename = ?", filename).first
+  if packet_log
+    Cosmos::Logger.warn("PacketLog already exists in database: #{filename}")
+  end
+
+  # Determine if this is a command or telemetry packet log
+  plr = Cosmos::PacketLogReader.new
+  plr.open(filename)
+  if plr.log_type == :TLM
+    is_tlm = true
+  else
+    is_tlm = false
+  end
+
+  build_lookups()
+
+  # Check if first and last packet in the log are already in the database
```

Comment:
First and last packets could already be in there because DART makes it own logs, and if you tried to import one of the files directly from the CmdTlmServer this could happen.   The first packet is SYSTEM META, which is a good point, as that wasn't true when I first wrote this.  It should probably check the first few packets and not just the first.   It is also totally possible that only the middle set of packets in the log file is already in DART...   A better check is probably to just check the time range of the packets in the file and see if DART already has data for that time range.  I will reconsider this.

Reply:
How can you import one of the files directly from the CmdTlmServer?

Reply:
So say DART wasn't running and then you want to import some data from the regular CmdTlmServer into DART.   You copy the regular COSMOS log file to System.paths['DART_DATA'] and then you run dart_import to get it into DART.   All the import does is create PacketLogEntry rows in the database.  The raw data is stored as is, where is, in the file being imported.  The file is now a part of the DART "database".